TrailBlaze/Networking/Interceptor.m

```diff
@@ -0,0 +1,125 @@
+    float shortestDistance = FLT_MAX;
+    int index = 0;
+    int closestIndex = -1;
+//    CLLocation* runnerCLLocation = [[CLLocation alloc] initWithLatitude:[[runnerLocation firstObject] doubleValue] longitude:[[runnerLocation lastObject] doubleValue]];
+    for (NSArray* point in routePoints) {
+        CLLocation* thisPointCLLocation = [[CLLocation alloc] initWithLatitude:[[point firstObject] doubleValue] longitude:[[point lastObject] doubleValue]];
+        float distanceBetween = [runnerLocation distanceFromLocation:thisPointCLLocation];
+        if ( distanceBetween < shortestDistance) {
+            shortestDistance = distanceBetween;
+            closestIndex = index;
+        }
+        index += 1;
+    }
+    return [routePoints subarrayWithRange:NSMakeRange(closestIndex, routePoints.count-closestIndex)];
+}
+
++ (NSArray *) sortInAscendingOrder: (NSArray *) etaDifferences {
+    NSSortDescriptor *sortDescriptor;
+    sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"doubleValue" ascending:YES];
+    NSArray *sortDescriptors = [NSArray arrayWithObject:sortDescriptor];
+    return [etaDifferences sortedArrayUsingDescriptors:sortDescriptors];
+}
+
++ (NSArray *) prunePoints: (NSArray *) routePoints numberOfPoints: (int) numberOfPoints{
+    NSMutableArray *prunedPoints = [[NSMutableArray alloc] init];
+    NSUInteger gap = routePoints.count/numberOfPoints;
+    for (int i = (int) gap; i < numberOfPoints; i+=gap) {
+        [prunedPoints addObject:[routePoints objectAtIndex:i]];
+    }
+    return prunedPoints;
+}
+
++ (void) getBestETAPoint: (NSArray *) allPoints interceptorLocation: (CLLocation *) interceptorLocation runnerLocation: (CLLocation *) runnerLocation completion:(void (^)(MKMapItem *bestPoint, NSError *))completion {
+    
+    NSArray *prunedPoints = [self prunePoints:[self getRemainingRoutePoints:allPoints runnerLocation:runnerLocation] numberOfPoints:10];
+    NSLog(@"🌈🌈🌈🌈🌈%@", prunedPoints);
+    etasDifferences = [NSMutableArray array];
+    etaPointPairs = [[NSMutableDictionary alloc] init];
+    dispatch_queue_t serialQueue = dispatch_queue_create("serialQueue", nil);
+    dispatch_group_t group = dispatch_group_create();
+    
+    MKPlacemark *startInterceptorPlacemark = [[MKPlacemark alloc] initWithCoordinate:CLLocationCoordinate2DMake(interceptorLocation.coordinate.latitude, interceptorLocation.coordinate.longitude)];
+    
+    MKPlacemark *startRunnerPlacemark = [[MKPlacemark alloc] initWithCoordinate:CLLocationCoordinate2DMake(runnerLocation.coordinate.latitude, runnerLocation.coordinate.longitude)];
+    
+    MKMapItem *startInterceptorItem = [[MKMapItem alloc] initWithPlacemark:startInterceptorPlacemark];
+    
+    MKMapItem *startRunnerItem = [[MKMapItem alloc] initWithPlacemark:startRunnerPlacemark];
+
+    for (int i=0; i < prunedPoints.count; i++) {
+        NSLog(@"starting operation %@..", @(i));
+        
+        dispatch_group_enter(group);
```

Comment:
i don't think it's functionally different, but i like to put `group_enter` _right_ before the async call (so right above line 91) so that as you read you can see it's being added to a dispatch group.